### PR TITLE
WT-9974 Enforce C and C++ style commenting in WiredTiger

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

**Comment by luke-pearson** [comment]
This is a draft PR, currently it is picking up 51 illegal comments which will need to be fixed in order for s_all to pass.

File: src/reconcile/rec_col.c
Comment (luke-pearson) [inline src/reconcile/rec_col.c:11]: I'm following up with @ajmorton as to whether these // 9417 ignore comments are meaningful.

Comment (ajmorton) [reply src/reconcile/rec_col.c:11]: No we can drop them. 
These were notes to myself when doing WT-9417 that should have been removed and slipped through review. There's a second one in `rec_child.c` that can also go

File: dist/comment_style.py
Comment (luke-pearson) [inline dist/comment_style.py:121]: When I originally wrote this I think i found a number of false positives in these directories and ignored them. I tried un-ignoring them and running the script and there is a tiny number of failures. We should remove this array and fix those failures if we can.

Comment (korteland) [reply dist/comment_style.py:121]: Ack, I'll give it a shot.

Comment (korteland) [reply dist/comment_style.py]: Done, `mtx_rw.c` had a bunch of embedded comments so I've left that in the list and fixed the others.

File: bench/workgen/workgen.cxx
Comment (ajmorton) [inline bench/workgen/workgen.cxx:111]: Sidenote: Does anyone know if there's a WT ticket to change the last of our `.cxx` files to `.cpp`?

File: dist/comment_style.py
Comment (ajmorton) [inline dist/comment_style.py:5]: nit: Can we rename to something like `report_illegal_comment`?

File: dist/comment_style.py
Comment (ajmorton) [inline dist/comment_style.py:6]: I'd recommend an f-string here:
`f"Illegal {multiline} comment in {file_name}:{line_num - 1} {line.strip()}"`
`.strip` also trims all whitespace by default so I don't think we need the `\n` arg here

File: dist/comment_style.py
Comment (ajmorton) [inline dist/comment_style.py:23]: I'd add a comment here explaining that `//` as part of a URL is a false positive we want to ignore.

File: dist/comment_style.py
Comment (ajmorton) [inline dist/comment_style.py:26]: A full explanation is gonna be tough, but we can we some details as to what cases we're catching here?
Something like:
```
Detect all occurences of `//` on a line. This can be at the start of a line, or after preceding text
```

I'm not sure exactly what case we're handling in this line so I'll leave that to you

Comment (ajmorton) [reply dist/comment_style.py:26]: Also as we're not making use of captured groups we can drop the `?:`s here.

File: dist/comment_style.py
Comment (ajmorton) [inline dist/comment_style.py:21]: Proposed alternate name: `detect_single_line_comments` and then a comment clarifying we're looking for any occurrences of a `//` comment in the file

File: dist/comment_style.py
Comment (ajmorton) [inline dist/comment_style.py:66]: We don't have any `.hpp` files so we can drop that extension

File: dist/comment_style.py
Comment (ajmorton) [inline dist/comment_style.py:112]: nit: we should be able to use 
```
for line, expected_result in tests:
    validate(line, expected_result)
``` 
for easier reading

Comment (ajmorton) [reply dist/comment_style.py:112]: The `sys.exit(0)` call also need to be de-indented so it's outside the for loop

File: dist/comment_style.py
Comment (ajmorton) [inline dist/comment_style.py:120]: It looks like the `ignored_folders` list is now gone. Do we no longer need it?

File: src/config/config.c
Comment (ajmorton) [inline src/config/config.c:102]: Where did this comment come from? It doesn't seem like part of the PR

File: src/include/log.h
Comment (ajmorton) [inline src/include/log.h:368]: nit: The FIXME should start on a new line

File: bench/workgen/workgen.cxx
Comment (luke-pearson) [reply bench/workgen/workgen.cxx:111]: I'll do it after we merge this change.

File: dist/comment_style.py
Comment (luke-pearson) [reply dist/comment_style.py:120]: We found that we could fix most occurrences in the {{ignored_folders}} list. Its also a bit misleading as technically we only search a subset of the folders to begin with. i.e. `bench examples ext src test`

File: src/include/log.h
Comment (korteland) [reply src/include/log.h:368]: Alas, `s_all` disagrees :-(

File: src/config/config.c
Comment (korteland) [reply src/config/config.c:102]: Yeah, I was talking to Michael about this code and it turns out there used to be a comment giving credit, but it went missing at some point. So I figured while I was here I'd reinstate it.

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py:120]: Nope - this came out of Luke's comment here: https://github.com/wiredtiger/wiredtiger/pull/8629#discussion_r1061045551

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py:66]: I get it, but I'd prefer to leave it in. It's a reasonable choice for future code to add, and leaving it out will cause hard-to-spot issues. There's also no real cost to its presence here.

File: src/include/log.h
Comment (ajmorton) [reply src/include/log.h:368]: hmm. Bit surprising. I think there's two way to fix that:
1. Add a second newline
```
* WiredTiger release version where log format version changed. 
* 
* FIXME WT-8681 - A
```

2. Add some leading whitespace
```
* WiredTiger release version where log format version changed. 
*   FIXME WT-8681 - A
```
Can we see if one of those fixes things?

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py]: Yep - done in bf9d525.

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py:6]: I don't think we can do that, since f-strings are a Python 3.6 addition and I think at least one of the builders is 3.4.

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py:24]: Thanks, done in bf9d525.

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py]: I reviewed the regex and realised a part of it was redundant. I've added a comment in bf9d525 that also addresses your `?:` comment.

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py]: Done in bf9d525, thanks.

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py]: Done in bf9d525, cheers for that.

File: src/include/log.h
Comment (korteland) [reply src/include/log.h]: The first one worked, so I've gone with that (3ea7742). Out of curiosity, why the need to have it on a new line?

Comment (ajmorton) [reply src/include/log.h]: For me it's just easier to see there's a fixme present in the code when reading. 
Purely cosmetic, hence the nit.

File: dist/comment_style.py
Comment (ajmorton) [reply dist/comment_style.py:6]: No worries. Our [min python version](https://wiki.corp.mongodb.com/display/WT/Tool%2C+compiler+and+language+standard+versions) is (in theory) 3.7 but happy to drop it if we know there are issues.

**Review by ajmorton — APPROVED** [review]
LGTM. Thanks for all the changes!

**Comment by luke-pearson** [comment]
This looks to me.

File: dist/comment_style.py
Comment (korteland) [reply dist/comment_style.py:6]: Yeah I did it for another PR last year and broke a pile of stuff unfortunately.

**Review by jeremythorp — APPROVED** [review]
LGTM!

**Comment by korteland** [comment]
evergreen merge

**Pull request merged**
